lib/Transforms/Scalar/PromoteMemcpy.cc

```diff
@@ -0,0 +1,211 @@
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/BuildLibCalls.h"
+#include "llvm/Transforms/Utils/Local.h"
+#include "llvm/Transforms/Utils/SimplifyLibCalls.h"
+
+#include "avy/AvyDebug.h"
+
+#define PMCPY_LOG(...) LOG("promote-memcpy", __VA_ARGS__)
+
+using namespace llvm;
+
+namespace {
+class PromoteMemcpy : public FunctionPass {
+public:
+  static char ID;
+
+  PromoteMemcpy() : FunctionPass(ID) {}
+
+  bool runOnFunction(Function &F);
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<llvm::DominatorTreeWrapperPass>();
+    AU.addRequired<AssumptionCacheTracker>();
+  }
+
+  const char *getPassName() const override { return "PromoteMemcpy"; }
+
+private:
+  Module *m_M = nullptr;
+  LLVMContext *m_Ctx = nullptr;
+  const DataLayout *m_DL = nullptr;
+  DominatorTree *m_DT = nullptr;
+  AssumptionCache *m_AC = nullptr;
+
+  bool simplifyMemCpy(MemCpyInst *MCpy);
+};
+
+char PromoteMemcpy::ID = 0;
+
+bool PromoteMemcpy::simplifyMemCpy(MemCpyInst *MI) {
+  assert(MI);
+
+  unsigned DstAlign =
+      getKnownAlignment(MI->getArgOperand(0), *m_DL, MI, m_AC, m_DT);
+  unsigned SrcAlign =
+      getKnownAlignment(MI->getArgOperand(1), *m_DL, MI, m_AC, m_DT);
+  unsigned MinAlign = std::min(DstAlign, SrcAlign);
+  unsigned CopyAlign = MI->getAlignment();
+
+  if (CopyAlign < MinAlign) {
+    MI->setAlignment(ConstantInt::get(MI->getAlignmentType(), MinAlign, false));
+    return MI;
+  }
+
+  // If MemCpyInst length is 1/2/4/8 bytes then replace memcpy with
+  // load/store.
+  ConstantInt *MemOpLength = dyn_cast<ConstantInt>(MI->getArgOperand(2));
+  if (!MemOpLength)
+    return false;
+
+  // Source and destination pointer types are always "i8*" for intrinsic.  See
+  // if the size is something we can handle with a single primitive load/store.
+  // A single load+store correctly handles overlapping memory in the memmove
+  // case.
+  uint64_t Size = MemOpLength->getLimitedValue();
+  if (Size == 0)
+    return false;
+
+  auto *SrcPtr = MI->getArgOperand(1);
+  auto *DstPtr = MI->getArgOperand(0);
+
+  unsigned SrcAddrSp = cast<PointerType>(SrcPtr->getType())->getAddressSpace();
+  unsigned DstAddrSp = cast<PointerType>(DstPtr->getType())->getAddressSpace();
+
+  if (SrcAddrSp != DstAddrSp)
+    return false;
+
+  auto *SrcBitCast = dyn_cast<BitCastInst>(SrcPtr);
+  auto *DstBitCast = dyn_cast<BitCastInst>(DstPtr);
+
+  if (!SrcBitCast || !DstBitCast)
+    return false;
+
+  auto *Src = SrcBitCast->getOperand(0);
+  auto *Dst = DstBitCast->getOperand(0);
+
+  auto *SrcPtrTy = dyn_cast<PointerType>(Src->getType());
+  assert(SrcPtrTy);
+
+  auto *DstPtrTy = dyn_cast<PointerType>(Dst->getType());
+  assert(DstPtrTy);
+
+  if (SrcPtrTy != DstPtrTy)
+    return false;
+
+  auto *Buffer = dyn_cast<AllocaInst>(Dst);
```

Comment:
Why does it matter that `memcpy` is into `alloca`? We want to minimize the number of memory operations.